### PR TITLE
Fix typing for `pl.overrides.distributed`

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7037 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

File: pytorch_lightning/overrides/distributed.py [hunk 1]
```diff
@@ -58,27 +59,27 @@ def _find_tensors(obj):  # pragma: no-cover
...
 # Note: Keep track of Pytorch DDP and update if there is a change
 # https://github.com/pytorch/pytorch/blob/v1.7.1/torch/nn/parallel/distributed.py#L626-L638
-def prepare_for_backward(model: DistributedDataParallel, output: Any):
+def prepare_for_backward(model: DistributedDataParallel, output: Any) -> None:
     # `prepare_for_backward` is `DistributedDataParallel` specific.
     if not isinstance(model, DistributedDataParallel):
         return
     if torch.is_grad_enabled() and model.require_backward_grad_sync:
-        model.require_forward_param_sync = True
+        model.require_forward_param_sync = True  # type: ignore[assignment]
```
Comment (Borda) [inline pytorch_lightning/overrides/distributed.py:67]: shall we define it as a union so allow bool?

File: pytorch_lightning/overrides/distributed.py [hunk 2]
```diff
@@ -91,13 +92,16 @@ class UnrepeatedDistributedSampler(DistributedSampler):
...
+        if not isinstance(self.dataset, Sized):
+            raise TypeError("The given dataset must implement the `__len__` method.")
         self.num_samples = len(range(self.rank, len(self.dataset), self.num_replicas))
         self.total_size = len(self.dataset)
         # If any process has at least one batch, every other process needs to
         # have at least one batch, or the DistributedDataParallel could lock up.
         assert self.num_samples >= 1 or self.total_size == 0
 
     def __iter__(self) -> Iterator[List[int]]:
+        assert isinstance(self.dataset, Sized)
```
Comment (Borda) [inline pytorch_lightning/overrides/distributed.py:104]: rather also if?

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/distributed.py:67]: this is an attribute defined by PyTorch nn.Module. Given the mypy error I see, I believe they have incorrect typing for this. I don't know how I would change the type like you said.

File: pytorch_lightning/overrides/distributed.py [hunk 3]
```diff
@@ -58,27 +59,27 @@ def _find_tensors(obj):  # pragma: no-cover
...
         # object. We need to find any tensors in this object, though,
         # because we need to figure out which parameters were used during
         # this forward pass, to ensure we short circuit reduction for any
         # unused parameters. Only if `find_unused_parameters` is set.
         if model.find_unused_parameters:
-            model.reducer.prepare_for_backward(list(_find_tensors(output)))
+            model.reducer.prepare_for_backward(list(_find_tensors(output)))  # type: ignore
         else:
-            model.reducer.prepare_for_backward([])
+            model.reducer.prepare_for_backward([])  # type: ignore
```
Comment (carmocca) [inline pytorch_lightning/overrides/distributed.py:76]: We can do this instead

```python
        args = list(_find_tensors(output)) if model.find_unused_parameters else []
        reducer = cast(Reducer, model.reducer)
        reducer.prepare_for_backward(args)
```

and avoid the ignores

File: pytorch_lightning/overrides/distributed.py [hunk 4]
```diff
@@ -42,11 +43,11 @@ def __init__(self, pl_module: "pl.LightningModule") -> None:
...
 
 
-def _find_tensors(obj):  # pragma: no-cover
-    r"""
-    Recursively find all tensors contained in the specified object.
-    """
-    if isinstance(obj, torch.Tensor):
+def _find_tensors(
+    obj: Union[Tensor, list, tuple, dict, Any]
+) -> Union[List[Tensor], itertools.chain]:  # pragma: no-cover
```
Comment (carmocca) [inline pytorch_lightning/overrides/distributed.py:48]: IMO this function could wrap with list the `itertools.chains` instead of having the callers do it.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/distributed.py:49]: maybe the implementation is this way to avoid a list conversion in the recursive calls

Comment (carmocca) [reply pytorch_lightning/overrides/distributed.py:48]: It could provide a base recursive case that does it.

But it's fine, it's like this in torch itself. Resolving
